tests/lib/modem_trace/modem_trace/src/modem_trace_test.c

```diff
@@ -0,0 +1,312 @@
+					   sizeof(sample_trace_data));
+	trace_processed_callback_ExpectAndReturn(sample_trace_data, sizeof(sample_trace_data), 0);
+
+	/* Let trace thread run. */
+	k_sleep(K_MSEC(1));
+}
+
+/* With the module correctly initialized, test that traces are forwarded to the trace medium
+ * even when nrf_modem_lib_trace_start has not yet been called.
+ */
+void test_modem_trace_process_when_modem_trace_start_was_not_called(void)
+{
+	int ret;
+	const uint8_t sample_trace_data[10];
+
+	__wrap_trace_medium_init_ExpectAndReturn(0);
+	TEST_ASSERT_EQUAL(0, nrf_modem_lib_trace_init(&trace_heap));
+
+	/* Simulate reception of a modem trace. */
+	ret = nrf_modem_lib_trace_process(sample_trace_data, sizeof(sample_trace_data));
+	TEST_ASSERT_EQUAL(0, ret);
+
+	__wrap_trace_medium_write_ExpectAndReturn(sample_trace_data, sizeof(sample_trace_data),
+					   sizeof(sample_trace_data));
+	trace_processed_callback_ExpectAndReturn(sample_trace_data, sizeof(sample_trace_data), 0);
+
+	/* Let trace thread run. */
+	k_sleep(K_MSEC(1));
+}
+
+/* Test that trace_medium_write is not called when trace module fails to initialize, and that
+ * nrf_modem_trace_processed_callback is called even when initialization fails.
+ */
+void test_modem_trace_write_when_medium_init_fails(void)
+{
+	int ret;
+	const uint8_t sample_trace_data[10];
+
+	__wrap_trace_medium_init_ExpectAndReturn(-EBUSY);
+	TEST_ASSERT_EQUAL(-EBUSY, nrf_modem_lib_trace_init(&trace_heap));
+	TEST_ASSERT_EQUAL(-ENXIO, nrf_modem_lib_trace_start(NRF_MODEM_LIB_TRACE_ALL));
+
+	/* Verify that nrf_modem_trace_processed_callback is called even when initialization
+	 * of trace module failed.
+	 */
+	trace_processed_callback_ExpectAndReturn(sample_trace_data, sizeof(sample_trace_data), 0);
+
+	/* Simulate the reception of a modem trace and expect no calls to trace_medium_write. */
+	ret = nrf_modem_lib_trace_process(sample_trace_data, sizeof(sample_trace_data));
+	TEST_ASSERT_EQUAL(-ENXIO, ret);
+}
+
+/* Test that trace_medium_write is not called when trace module has not been initialized, and that
+ * nrf_modem_trace_processed_callback is called even when trace module has not been initialized.
+ */
+void test_modem_trace_process_when_not_initialized(void)
+{
+	int ret;
+	const uint8_t sample_trace_data[10];
+
+	/* Verify that nrf_modem_trace_processed_callback is called even when the trace module
+	 * has not been initialized.
+	 */
+	trace_processed_callback_ExpectAndReturn(sample_trace_data, sizeof(sample_trace_data), 0);
+
+	/* Simulate the reception of a modem trace and expect no calls to trace_medium_write. */
+	ret = nrf_modem_lib_trace_process(sample_trace_data, sizeof(sample_trace_data));
+	TEST_ASSERT_EQUAL(-ENXIO, ret);
+}
+
+/* Test nrf_modem_lib_trace_start when nrf_modem_at_printf returns fails. */
+void test_modem_trace_start_when_nrf_modem_at_printf_fails(void)
+{
+	__wrap_trace_medium_init_ExpectAndReturn(0);
+	TEST_ASSERT_EQUAL(0, nrf_modem_lib_trace_init(&trace_heap));
```

Comment:
Here, and other places, you could reuse `test_modem_trace_init_medium`. If the init sequence is updated, it will be fewer places to do so.